### ISSUE TITLE
[Question] About micro batches schedule

### ISSUE THREAD

**Issue by lviy** [THUDM/slime#1964]
### Your Question

Hi Slime team, I have a question about the logprob computation and training setup. Why do logprob computation and training need to use the same micro-batch schedule? Is this because there is evidence or research suggesting that different micro-batch sizes may lead to accuracy issues? Also, would the team consider implementing --log-probs-max-tokens-per-gpu? 

I find peak GPU memory usage in logprob computation is lower than train stage, so I implement the --log-probs-max-tokens-per-gpu to reduce the number of micro batches to accelerate logprob computation stage. 

<img width="1202" height="744" alt="Image" src="https://github.com/user-attachments/assets/df07b94b-6a76-4525-818a-a93ca53e3ac5" />  
This figure shows the difference in the training arguments.

### What I've Tried

I read the source code of dynamic batching

### Environment (if relevant)

- slime version: latest
- Python version:
- PyTorch version:
- CUDA/ROCm version:
- GPU type and count:
- OS:

### Pre-submission Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/THUDM/slime/blob/main/CONTRIBUTING.md) and understand the collaboration scope.
- [x] I have read the [documentation](https://thudm.github.io/slime/) and [FAQ](https://thudm.github.io/slime/en/get_started/qa.html) and my question is not answered there.
- [x] I have searched for [existing issues](https://github.com/THUDM/slime/issues) and my question has not been asked before.

**Comment by zhuzilin** [comment]
Thanks for raising this question.

The main reason we currently keep the micro-batch schedule the same between logprob computation and training is to minimize the numerical difference between the logprobs computed before training and the logprobs recomputed during training.

In principle, changing the micro-batch size should not change the mathematical result. However, in practice, especially with large-scale distributed training, different micro-batch schedules may lead to slightly different execution paths, padding/batching behavior, kernel choices, reduction orders, etc. These differences can introduce small numerical discrepancies. For RL training, we found it safer to keep the two schedules aligned so that the recomputed logprobs are as close as possible to the original logprobs.

Actually, our first implementation did include a separate --log-probs-max-tokens-per-gpu option for the logprob stage, because, as you mentioned, logprob computation usually has lower peak memory usage than the training stage and can therefore use larger micro-batches for better efficiency. However, we later observed some logprob differences caused by using different schedules. Around that time, recomputation-related numerical errors were also receiving a lot of attention in the community, so we decided to make the logprob and training schedules consistent by default.

So this is mainly a stability/correctness-oriented design choice rather than a strict theoretical requirement. We agree that a separate --log-probs-max-tokens-per-gpu can improve logprob-stage efficiency, and it may still be useful as an opt-in feature. But if we add it back, we would probably want to make the potential numerical difference explicit, and ideally provide some validation or warning mechanism to help users check the logprob diff in their own setting.

Thanks again for pointing this out. This is definitely a useful trade-off to discuss.

**Comment by lviy** [comment]
> Thanks for raising this question.
> 
> The main reason we currently keep the micro-batch schedule the same between logprob computation and training is to minimize the numerical difference between the logprobs computed before training and the logprobs recomputed during training.
> 
> In principle, changing the micro-batch size should not change the mathematical result. However, in practice, especially with large-scale distributed training, different micro-batch schedules may lead to slightly different execution paths, padding/batching behavior, kernel choices, reduction orders, etc. These differences can introduce small numerical discrepancies. For RL training, we found it safer to keep the two schedules aligned so that the recomputed logprobs are as close as possible to the original logprobs.
> 
> Actually, our first implementation did include a separate --log-probs-max-tokens-per-gpu option for the logprob stage, because, as you mentioned, logprob computation usually has lower peak memory usage than the training stage and can therefore use larger micro-batches for better efficiency. However, we later observed some logprob differences caused by using different schedules. Around that time, recomputation-related numerical errors were also receiving a lot of attention in the community, so we decided to make the logprob and training schedules consistent by default.
> 
> So this is mainly a stability/correctness-oriented design choice rather than a strict theoretical requirement. We agree that a separate --log-probs-max-tokens-per-gpu can improve logprob-stage efficiency, and it may still be useful as an opt-in feature. But if we add it back, we would probably want to make the potential numerical difference explicit, and ideally provide some validation or warning mechanism to help users check the logprob diff in their own setting.
> 
> Thanks again for pointing this out. This is definitely a useful trade-off to discuss.

Thanks for your response! I learned a lot from it.
